exchange_calendars/exchange_calendar_xkrx.py

```diff
@@ -569,18 +64,195 @@ class XKRXExchangeCalendar(PrecomputedExchangeCalendar):
+                    self.name,
+                    self.first_trading_session.year,
+                ),
+            )
 
-    def __init__(self, *args, **kwargs):
-        super(XKRXExchangeCalendar, self).__init__(*args, **kwargs)
+        if latest_precomputed_year < self.last_trading_session.year:
+            raise ValueError(
+                "The {} holidays are only recorded to {},"
+                " cannot instantiate the {} calendar for {}.".format(
+                    self.name,
+                    latest_precomputed_year,
+                    self.name,
+                    self.last_trading_session.year,
+                ),
+            )
+
+    # KRX schedule change history
+    # https://blog.naver.com/daishin_blog/220724111002
+
+    # 1956-03-03: 0930~1130, 1330~1530
+    # 1978-04-??: 1000~1200, 1330~1530
+    # 1986-04-??: 0940~1200, 1320~1520
+    # 1987-03-??: 0940~1140, 1320~1520
+    # 1995-01-01: 0930~1130, 1300~1500
+    # 1998-12-07: 0900~1200, 1300~1500
+    # 2000-05-22: 0900~1500
+    # 2016-08-01: 0900~1530
+
+    # Break time disappears since 2000-05-22
+    # https://www.donga.com/news/Economy/article/all/20000512/7534650/1
+
+    # Closing time became 30mins late since 2016-08-01
+    # https://biz.chosun.com/site/data/html_dir/2016/07/24/2016072400309.html
+
+    open_times = (
+        (None, time(9, 30)),
+        (pd.Timestamp("1978-04-01"), time(10, 0)),
+        (pd.Timestamp("1986-04-01"), time(9, 40)),
+        (pd.Timestamp("1995-01-01"), time(9, 30)),
+        (pd.Timestamp("1998-12-07"), time(9, 0)),
+    )
+    break_start_times = (
+        (None, time(11, 30)),
+        (pd.Timestamp("1978-04-01"), time(12, 0)),
+        (pd.Timestamp("1987-03-01"), time(11, 40)),
+        (pd.Timestamp("1995-01-01"), time(11, 30)),
+        (pd.Timestamp("1998-12-07"), time(12, 0)),
+        (pd.Timestamp("2000-05-22"), None),
+    )
+    break_end_times = (
+        (None, time(13, 30)),
+        (pd.Timestamp("1986-04-01"), time(13, 20)),
+        (pd.Timestamp("1995-01-01"), time(13, 0)),
+        (pd.Timestamp("2000-05-22"), None),
+    )
+    close_times = (
+        (None, time(15, 30)),
+        (pd.Timestamp("1986-04-01"), time(15, 20)),
+        (pd.Timestamp("1995-01-01"), time(15, 0)),
+        (pd.Timestamp("2016-08-01"), time(15, 30)),
+    )
+
+    # Saterday became holiday since 1998-12-07
+    # https://www.hankyung.com/finance/article/1998080301961
+
+    weekmask = "1111100"
 
     @property
-    def name(self):
-        # Korea Exchange
-        return "XKRX"
+    def special_weekmasks(self):
+        return [
+            (None, pd.Timestamp("1998-12-07") - pd.Timedelta(1, unit="D"), "1111110"),
+        ]
+
+    # KRX regular and adhoc holidays
+
+    @property
+    def regular_holidays(self):
+        return HolidayCalendar(krx_regular_holiday_rules)
+
+    @property
+    def adhoc_holidays(self):
+        return precomputed_krx_holidays.tolist()
```

Comment:
What's the overlap between the precomputed holiday and the holidays generated by the new rules? Or are they disjoint?

Reply:
Since the original precomputed holidays covered both regular and non-regular (adhoc) holidays when the calendar was `PrecomputedExchangeCalendar`, each of newly suggested regular holidays (generated by the new rules) should generally fall into one of the original precomputed holidays (except the future holidays that are not precomputed yet).

So they are not disjoint. Possibly the generated holidays should be a subset of precomputed holidays.

Things to note however:

1. precomputed holidays pre-emptively exclude weekends (saturdays and sundays) while the generated holidays do not.
2. precomputed holidays may include adhoc holidays such as election day while the generated holidays cannot cover them.
3. still precomputed holidays may omit some valid holidays and generated holidays might cover those missing holidays.
4. and still there might be some valid adhoc holidays that both precomputed and generated holidays miss out.

After listening to your question, I quickly added a test to check whether my argument is right (with additional fix to labor day). Currently that test fails but I think those cases all fall into the case 3 explained above.

<details><summary>Generated holidays that don't fall into the precomputed holidays</summary>
<p>

```
>       self.assertTrue(all(isin), "missing holidays = \n%s" % missing)
E       AssertionError: False is not true : missing holidays =
E       1975-09-19        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1976-01-30    Seollal (New Year's Day by the lunar) (-1 day)
E       1976-09-07        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1976-09-09        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1977-02-17    Seollal (New Year's Day by the lunar) (-1 day)
E       1977-02-18             Seollal (New Year's Day by the lunar)
E       1977-09-26        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1977-09-28        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1978-02-06    Seollal (New Year's Day by the lunar) (-1 day)
E       1978-02-07             Seollal (New Year's Day by the lunar)
E       1978-02-08    Seollal (New Year's Day by the lunar) (+1 day)
E       1978-09-18        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1979-01-29    Seollal (New Year's Day by the lunar) (+1 day)
E       1979-10-04        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1980-02-15    Seollal (New Year's Day by the lunar) (-1 day)
E       1980-09-22        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1980-09-24        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1981-02-04    Seollal (New Year's Day by the lunar) (-1 day)
E       1981-02-05             Seollal (New Year's Day by the lunar)
E       1981-02-06    Seollal (New Year's Day by the lunar) (+1 day)
E       1981-09-11        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1982-01-25             Seollal (New Year's Day by the lunar)
E       1982-01-26    Seollal (New Year's Day by the lunar) (+1 day)
E       1982-09-30        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1983-02-14    Seollal (New Year's Day by the lunar) (+1 day)
E       1983-09-20        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1983-09-22        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1984-02-01    Seollal (New Year's Day by the lunar) (-1 day)
E       1984-02-02             Seollal (New Year's Day by the lunar)
E       1984-02-03    Seollal (New Year's Day by the lunar) (+1 day)
E       1984-09-11        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1985-02-19    Seollal (New Year's Day by the lunar) (-1 day)
E       1985-02-21    Seollal (New Year's Day by the lunar) (+1 day)
E       1985-09-30        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1986-02-10    Seollal (New Year's Day by the lunar) (+1 day)
E       1986-09-17        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1987-01-28    Seollal (New Year's Day by the lunar) (-1 day)
E       1987-01-30    Seollal (New Year's Day by the lunar) (+1 day)
E       1987-10-06        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1988-02-17    Seollal (New Year's Day by the lunar) (-1 day)
E       1988-02-19    Seollal (New Year's Day by the lunar) (+1 day)
E       1996-02-19             Seollal (New Year's Day by the lunar)
E       1996-02-20    Seollal (New Year's Day by the lunar) (+1 day)
E       1997-05-01                                         Labor Day
E       1997-05-14                                 Buddha's Birthday
```

</p>
</details>

I'm thinking of putting those holidays to precomputed list or just limiting the testing date range not to include those dates, when I'm confident with my argument.

Some external references for that:

1. [1997-05-01 is holiday because it's Labor day](https://www.hankyung.com/finance/article/1997042802391)
2. [All the dates above are missing in historical daily quotes of a symbol](https://finance.daum.net/quotes/A000020#current/quote)
3. [1997-12-18 is holiday for election but both precomputed and generated are missing out this holiday](https://www.hankyung.com/finance/article/1997121600791)